common/rpc/interceptor/health_check.go

```diff
@@ -0,0 +1,127 @@
+package interceptor
+
+import (
+	"context"
+	"time"
+
+	"go.temporal.io/api/serviceerror"
+	"go.temporal.io/server/common"
+	"go.temporal.io/server/common/aggregate"
+	"go.temporal.io/server/common/log"
+	"google.golang.org/grpc"
+)
+
+// HealthCheckInterceptor is a gRPC interceptor that records health metrics
+type HealthCheckInterceptor struct {
+	historyHealthSignalAggregator HealthSignalAggregator
+}
+
+// NewHealthCheckInterceptor creates a new health check interceptor
+func NewHealthCheckInterceptor(historyHealthSignalAggregator HealthSignalAggregator) *HealthCheckInterceptor {
+	return &HealthCheckInterceptor{
+		historyHealthSignalAggregator: historyHealthSignalAggregator,
+	}
+}
+
+// UnaryIntercept implements the gRPC unary interceptor interface
+func (h *HealthCheckInterceptor) UnaryIntercept(
+	ctx context.Context,
+	req interface{},
+	info *grpc.UnaryServerInfo,
+	handler grpc.UnaryHandler,
+) (interface{}, error) {
+	startTime := time.Now()
+	resp, err := handler(ctx, req)
+	elapsed := time.Since(startTime)
+
+	h.historyHealthSignalAggregator.Record(elapsed, err)
+
+	return resp, err
+}
+
+type (
```

Comment:
Is it possible to make this its own file and share between the interceptor and persistence?

Reply:
where would be a good shared package? Mentioned in an earlier comment that I didn't do this on purpose for simplicity of the PR https://github.com/temporalio/temporal/pull/7773#discussion_r2093650649